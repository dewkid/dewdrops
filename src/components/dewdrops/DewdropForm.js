import React from 'react';
import PropTypes from 'prop-types';
import TextInput from '../common/TextInput';
import ImageDropZone from '../common/ImageDropZone';

const DewdropForm = ({
  dewdrop,
  onSave,
  onChange,
  onFilesDrop,
  saving = false,
  errors = {}
}) => {
  return (
    <form onSubmit={onSave}>
      <h2>{dewdrop.id ? 'Edit' : 'Add'} Dewdrop</h2>
      {errors.onSave && (
        <div className='alert alert-danger' role='alert'>
          {errors.onSave}
        </div>
      )}

      <div className='row'>
        <div className='col-md-4'>
          <ImageDropZone
            name='image'
            label='Image'
            onFilesDrop={onFilesDrop}
            error={errors.image}
          />
        </div>
      </div>
      <div className='row'>
        <div className='col-md-8'>
          <TextInput
            name='caption'
            label='Caption'
            value={dewdrop.caption}
            onChange={onChange}
            error={errors.caption}
          />
        </div>
      </div>

      <button type='submit' disabled={saving} className='btn btn-primary'>
        {saving ? 'Saving...' : 'Save'}
      </button>
    </form>
  );
};

DewdropForm.propTypes = {
  dewdrop: PropTypes.object.isRequired,
  errors: PropTypes.object,
  onSave: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onFilesDrop: PropTypes.func.isRequired,
  saving: PropTypes.bool
};

export default DewdropForm;
