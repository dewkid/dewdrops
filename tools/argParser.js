function setPassword(argv) {
  const args = argv.filter(value => value.startsWith('--password='));
  if (args.length == 1) {
    let passwordArg = args[0];
    process.env.DB_PASSWORD = passwordArg.substring(11);
  }
  if (!process.env.DB_PASSWORD || process.env.DB_PASSWORD.length == 0) {
    console.log('--password=xxxx argument is required');
    process.exit(1);
  }
}

module.exports = {
  setPassword
};
