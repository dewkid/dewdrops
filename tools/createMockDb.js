/* eslint-disable no-console */
const util = require('util');
const fs = require('fs');
const rimraf = require('rimraf');
const path = require('path');
const { dewdrops, comments, users } = require('./mockData');
const copyFilePromise = util.promisify(fs.copyFile);
const { setPassword } = require('./argParser');

setPassword(process.argv);

const { User, Dewdrop, Comment, connection } = require('./dbModels');

function copyFiles(srcDir, destDir, files) {
  return Promise.all(
    files.map(f => {
      return copyFilePromise(path.join(srcDir, f), path.join(destDir, f));
    })
  );
}

const imageDataPath = path.join(__dirname, 'imageData');
const mockImageDataPath = path.join(__dirname, 'mockImages');

// remove the image directory
rimraf.sync(imageDataPath);

// recreate image directory and copy mock images
if (!fs.existsSync(imageDataPath)) fs.mkdirSync(imageDataPath);
copyFiles(mockImageDataPath, imageDataPath, [
  'bob.jpg',
  'character.jpg',
  'ski.jpg',
  'sunset.jpg',
  'void.jpg'
])
  .then(() => {
    console.log('done');
    setupPostgresTables();
  })
  .catch(err => {
    console.log(err);
  });

// Testing

function testConnection() {
  connection
    .authenticate()
    .then(() => {
      console.log('Connection has been established successfully.');
    })
    .catch(err => {
      console.error('Unable to connect to the database:', err);
    });
}

async function createTableData() {
  await Promise.all(
    users.map(user => {
      delete user.id;
      User.create(user);
    })
  );
  await Promise.all(
    dewdrops.map(dewdrop => {
      delete dewdrop.id;
      Dewdrop.create(dewdrop);
    })
  );
  await Promise.all(
    comments.map(comment => {
      delete comment.id;
      Comment.create(comment);
    })
  );
}

function setupPostgresTables() {
  testConnection();
  connection.sync({ force: true }).then(() => {
    console.log('db tables created');
    createTableData().then(() => {
      console.log('Mock records created');

      Dewdrop.findAll().then(dewdrops => {
        dewdrops.map(dewdrop => console.log(dewdrop.imageName));
      });
    });
  });
}
