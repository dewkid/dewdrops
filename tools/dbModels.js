const Sequelize = require('sequelize');

const connection = new Sequelize(
  'dewdropsweb',
  'postgres',
  process.env.DB_PASSWORD,
  {
    host: 'localhost',
    dialect: 'postgres' /* one of 'mysql' | 'mariadb' | 'postgres' | 'mssql' */,
    logging: false
  }
);

const User = connection.define(
  'user',
  {
    // attributes
    name: {
      type: Sequelize.STRING,
      allowNull: false
    }
  },
  {
    // options
  }
);

const Dewdrop = connection.define(
  'dewdrop',
  {
    // attributes
    userid: {
      type: Sequelize.INTEGER,
      allowNull: false
    },
    imageName: {
      type: Sequelize.STRING,
      allowNull: false
    },
    caption: {
      type: Sequelize.STRING,
      allowNull: false
    }
  },
  {
    // options
  }
);

const Comment = connection.define(
  'comment',
  {
    // attributes
    dewdropid: {
      type: Sequelize.INTEGER,
      allowNull: false
    },
    userid: {
      type: Sequelize.INTEGER,
      allowNull: false
    },
    commentText: {
      type: Sequelize.STRING,
      allowNull: false
    }
  },
  {
    // options
  }
);

// Using CommonJS style export so we can consume via Node (without using Babel-node)
module.exports = {
  connection,
  User,
  Dewdrop,
  Comment
};
