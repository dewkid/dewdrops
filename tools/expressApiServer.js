const express = require('express');
const bodyParser = require('body-parser');
const fs = require('fs');
const formidable = require('formidable');

const baseImagePath = './tools/imageData/';
const server = express();

const { setPassword } = require('./argParser');

setPassword(process.argv);

const { Dewdrop, User, Comment } = require('./dbModels');

server.use(bodyParser.json());
server.use(express.static('app/public'));

server.use(function(req, res, next) {
  res.header('Access-Control-Allow-Origin', 'http://localhost:3000'); // the domain you will make the request from
  res.header(
    'Access-Control-Allow-Headers',
    'Origin, X-Requested-With, Content-Type, Accept'
  );
  next();
});

// Image uploads
server.post('/upload', function(req, res, next) {
  var form = new formidable.IncomingForm();
  form.parse(req, function(err, fields, files) {
    const file = files.file;
    const filename = fields.filename;

    // oldpath : temporary folder to which file is saved to
    var oldpath = file.path;
    var newpath = baseImagePath + filename;

    console.log('Copy from ' + oldpath + ' to ' + newpath);
    // copy the file to a new location (mv is probably a better choice)
    fs.copyFile(oldpath, newpath, function(err) {
      if (err) {
        res.status(400).send(err);
      } else {
        console.log('File uploaded and moved!');
        res.status(200).send('good');
        next();
      }
    });
  });
});

// Dewdrops

server.get('/dewdrops', (req, res) =>
  Dewdrop.findAll().then(result => res.json(result))
);

server.get('/dewdrops/:id', (req, res) =>
  Dewdrop.findByPk(req.params.id).then(result => {
    Dewdrop.max('id').then(max => {
      let dewdrop = result.toJSON();
      dewdrop.previd = dewdrop.id - 1;
      dewdrop.nextid = dewdrop.id < max ? dewdrop.id + 1 : 0;
      return res.json(dewdrop);
    });
  })
);

server.post('/dewdrops', (req, res) =>
  Dewdrop.create({ ...req.body }).then(result => res.json(result))
);

server.put('/dewdrops/:id', (req, res) =>
  Dewdrop.update(
    {
      ...req.body
    },
    {
      where: {
        id: req.params.id
      }
    }
  ).then(result => res.json(result))
);

server.delete('/dewdrops/:id', (req, res) =>
  Dewdrop.destroy({
    where: {
      id: req.params.id
    }
  }).then(result => res.json(result))
);

// Comments
server.get('/comments', (req, res) =>
  Comment.findAll().then(result => res.json(result))
);

server.get('/comments/:dewdropid', (req, res) =>
  // Actually loading all comments for a specified dewdrop
  Comment.findAll({
    where: { dewdropid: req.params.dewdropid }
  }).then(result => res.json(result))
);

server.post('/comments', (req, res) => {
  Comment.create({ ...req.body }).then(result => res.json(result));
});

// Users

server.get('/users', (req, res) =>
  User.findAll().then(result => res.json(result))
);

server.get('/users/:id', (req, res) =>
  User.findByPk(req.params.id).then(result => res.json(result))
);

server.listen(3001, () =>
  console.log('Express server listening on port 3001!')
);
