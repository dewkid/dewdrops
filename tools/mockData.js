const dewdrops = [
  {
    id: 1,
    userid: 1,
    imageName: 'character.jpg',
    caption: 'My next D&D character!'
  },
  {
    id: 2,
    userid: 2,
    imageName: 'sunset.jpg',
    caption: 'Check out this sunset'
  },
  {
    id: 3,
    userid: 1,
    imageName: 'ski.jpg',
    caption: 'Killer shot of me hitting the slopes.'
  },
  {
    id: 4,
    userid: 3,
    imageName: 'bob.jpg',
    caption: 'THIS guy!'
  }
];

const comments = [
  { id: 1, dewdropid: 1, userid: 2, commentText: 'Sweet!' },
  { id: 2, dewdropid: 1, userid: 3, commentText: "What's his name?" },
  {
    id: 3,
    dewdropid: 2,
    userid: 4,
    commentText: 'Where did you take that?  Is that near your place?'
  },
  { id: 4, dewdropid: 1, userid: 1, commentText: "I'm thinking Garr or Frank" },
  { id: 5, dewdropid: 1, userid: 3, commentText: 'Frank?' },
  { id: 6, dewdropid: 2, userid: 1, commentText: 'Yeah, where is that?' },
  { id: 7, dewdropid: 1, userid: 4, commentText: 'LOL - how about Jeeves' },
  { id: 8, dewdropid: 1, userid: 1, commentText: 'LOL' },
  { id: 9, dewdropid: 2, userid: 2, commentText: 'Not my pic, just liked it.' },
  {
    id: 10,
    dewdropid: 3,
    userid: 2,
    commentText: 'Right, like you could do that.'
  },
  { id: 11, dewdropid: 2, userid: 3, commentText: 'Lame!' },
  { id: 12, dewdropid: 3, userid: 1, commentText: "That's me!" },
  { id: 13, dewdropid: 2, userid: 1, commentText: 'I knew it!' },
  {
    id: 14,
    dewdropid: 4,
    userid: 4,
    commentText: 'Is that what you are wearing for the party?'
  },
  { id: 15, dewdropid: 4, userid: 3, commentText: 'Thinking about it...' },
  {
    id: 16,
    dewdropid: 4,
    userid: 3,
    commentText: 'You totally should, that would be hilarious.'
  }
];

const users = [
  { id: 1, name: 'Robin', username: 'robin', password: 'robin' },
  { id: 2, name: 'Dave', username: 'dave', password: 'dave' },
  { id: 3, name: 'Ross', username: 'ross', password: 'ross' },
  { id: 4, name: 'Ryan', username: 'ryan', password: 'ryan' },
  { id: 5, name: 'Lance', username: 'lance', password: 'lance' },
  { id: 6, name: 'Paul', username: 'paul', password: 'lance' },
  { id: 7, name: 'Nathaniel', username: 'nathaniel', password: 'nathaniel' },
  { id: 8, name: 'Chris', username: 'chris', password: 'chris' },
  { id: 9, name: 'Sam', username: 'sam', password: 'sam' }
];

const newDewdrop = {
  id: null,
  userid: null,
  imageName: '',
  caption: ''
};

const newComment = {
  id: null,
  dewdropid: null,
  userid: null,
  commentText: ''
};

// Using CommonJS style export so we can consume via Node (without using Babel-node)
module.exports = {
  newDewdrop,
  newComment,
  dewdrops,
  comments,
  users
};
