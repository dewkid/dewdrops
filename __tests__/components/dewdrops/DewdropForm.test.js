import React from 'react';
import renderer from 'react-test-renderer';
import { dewdrops } from '../../../tools/mockData';

// Component under test
import DewdropForm from '../../../src/components/dewdrops/DewdropForm';

it("sets submit button label 'Saving...' when saving is true", () => {
  const tree = renderer.create(
    <DewdropForm
      dewdrop={dewdrops[0]}
      onSave={jest.fn()}
      onChange={jest.fn()}
      onFilesDrop={jest.fn()}
      saving
      errors={{}}
    />
  );

  expect(tree).toMatchSnapshot();
});

it("sets submit button label 'Save' when saving is false", () => {
  const tree = renderer.create(
    <DewdropForm
      dewdrop={dewdrops[0]}
      onSave={jest.fn()}
      onChange={jest.fn()}
      onFilesDrop={jest.fn()}
      saving={false}
      errors={{}}
    />
  );

  expect(tree).toMatchSnapshot();
});

it('sets a caption field error when there is a caption error', () => {
  const tree = renderer.create(
    <DewdropForm
      dewdrop={{}}
      onSave={jest.fn()}
      onChange={jest.fn()}
      onFilesDrop={jest.fn()}
      saving={false}
      errors={{ caption: 'error' }}
    />
  );

  expect(tree).toMatchSnapshot();
});

it('sets an image field error when there is an image error', () => {
  const tree = renderer.create(
    <DewdropForm
      dewdrop={{}}
      onSave={jest.fn()}
      onChange={jest.fn()}
      onFilesDrop={jest.fn()}
      saving={false}
      errors={{ image: 'error' }}
    />
  );

  expect(tree).toMatchSnapshot();
});
