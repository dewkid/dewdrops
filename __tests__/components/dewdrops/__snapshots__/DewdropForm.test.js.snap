// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sets a caption field error when there is a caption error 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Add
     Dewdrop
  </h2>
  <div
    className="row"
  >
    <div
      className="col-md-4"
    >
      <div>
        <label
          htmlFor="image"
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Image
        </label>
        <div
          onBlur={[Function]}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fafafa",
              "borderColor": "#eeeeee",
              "borderRadius": 2,
              "borderStyle": "dashed",
              "borderWidth": 2,
              "color": "#bdbdbd",
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
              "height": "200px",
              "outline": "none",
              "padding": "20px",
              "transition": "border .24s ease-in-out",
            }
          }
          tabIndex={0}
        >
          <input
            accept="image/*"
            autoComplete="off"
            multiple={false}
            onChange={[Function]}
            onClick={[Function]}
            style={
              Object {
                "display": "none",
              }
            }
            tabIndex={-1}
            type="file"
          />
          <p>
            Drag-n-drop photo here, or click to select photo
          </p>
          <div />
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-8"
    >
      <div
        className="form-group has-error"
      >
        <label
          htmlFor="caption"
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Caption
        </label>
        <div
          className="field"
        >
          <input
            className="form-control"
            name="caption"
            onChange={[MockFunction]}
            type="text"
          />
          <div
            className="alert alert-danger"
          >
            error
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Save
  </button>
</form>
`;

exports[`sets an image field error when there is an image error 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Add
     Dewdrop
  </h2>
  <div
    className="row"
  >
    <div
      className="col-md-4"
    >
      <div>
        <label
          htmlFor="image"
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Image
        </label>
        <div
          onBlur={[Function]}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fafafa",
              "borderColor": "#eeeeee",
              "borderRadius": 2,
              "borderStyle": "dashed",
              "borderWidth": 2,
              "color": "#bdbdbd",
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
              "height": "200px",
              "outline": "none",
              "padding": "20px",
              "transition": "border .24s ease-in-out",
            }
          }
          tabIndex={0}
        >
          <input
            accept="image/*"
            autoComplete="off"
            multiple={false}
            onChange={[Function]}
            onClick={[Function]}
            style={
              Object {
                "display": "none",
              }
            }
            tabIndex={-1}
            type="file"
          />
          <p>
            Drag-n-drop photo here, or click to select photo
          </p>
          <div />
        </div>
        <div
          className="alert alert-danger"
        >
          error
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-8"
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="caption"
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Caption
        </label>
        <div
          className="field"
        >
          <input
            className="form-control"
            name="caption"
            onChange={[MockFunction]}
            type="text"
          />
        </div>
      </div>
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Save
  </button>
</form>
`;

exports[`sets submit button label 'Save' when saving is false 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit
     Dewdrop
  </h2>
  <div
    className="row"
  >
    <div
      className="col-md-4"
    >
      <div>
        <label
          htmlFor="image"
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Image
        </label>
        <div
          onBlur={[Function]}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fafafa",
              "borderColor": "#eeeeee",
              "borderRadius": 2,
              "borderStyle": "dashed",
              "borderWidth": 2,
              "color": "#bdbdbd",
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
              "height": "200px",
              "outline": "none",
              "padding": "20px",
              "transition": "border .24s ease-in-out",
            }
          }
          tabIndex={0}
        >
          <input
            accept="image/*"
            autoComplete="off"
            multiple={false}
            onChange={[Function]}
            onClick={[Function]}
            style={
              Object {
                "display": "none",
              }
            }
            tabIndex={-1}
            type="file"
          />
          <p>
            Drag-n-drop photo here, or click to select photo
          </p>
          <div />
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-8"
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="caption"
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Caption
        </label>
        <div
          className="field"
        >
          <input
            className="form-control"
            name="caption"
            onChange={[MockFunction]}
            type="text"
            value="My next D&D character!"
          />
        </div>
      </div>
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    type="submit"
  >
    Save
  </button>
</form>
`;

exports[`sets submit button label 'Saving...' when saving is true 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <h2>
    Edit
     Dewdrop
  </h2>
  <div
    className="row"
  >
    <div
      className="col-md-4"
    >
      <div>
        <label
          htmlFor="image"
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Image
        </label>
        <div
          onBlur={[Function]}
          onClick={[Function]}
          onDragEnter={[Function]}
          onDragLeave={[Function]}
          onDragOver={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fafafa",
              "borderColor": "#eeeeee",
              "borderRadius": 2,
              "borderStyle": "dashed",
              "borderWidth": 2,
              "color": "#bdbdbd",
              "display": "flex",
              "flex": 1,
              "flexDirection": "column",
              "height": "200px",
              "outline": "none",
              "padding": "20px",
              "transition": "border .24s ease-in-out",
            }
          }
          tabIndex={0}
        >
          <input
            accept="image/*"
            autoComplete="off"
            multiple={false}
            onChange={[Function]}
            onClick={[Function]}
            style={
              Object {
                "display": "none",
              }
            }
            tabIndex={-1}
            type="file"
          />
          <p>
            Drag-n-drop photo here, or click to select photo
          </p>
          <div />
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-8"
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="caption"
          style={
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "marginTop": 20,
            }
          }
        >
          Caption
        </label>
        <div
          className="field"
        >
          <input
            className="form-control"
            name="caption"
            onChange={[MockFunction]}
            type="text"
            value="My next D&D character!"
          />
        </div>
      </div>
    </div>
  </div>
  <button
    className="btn btn-primary"
    disabled={true}
    type="submit"
  >
    Saving...
  </button>
</form>
`;
